File: tutopy_pkg/tutopy_pkg/activity_04_client.py
#!/usr/bin/env python3
import rclpy
import rclpy.time

from functools import partial
from gonca_interfaces.srv import BatteryStatus
from rclpy.node import Node


class LEDClientNode(Node):
    def __init__(self):
        super().__init__("LED_client_node")
        self.led_to_set = 3
        self.battery_full = False
        self.last_time = self.get_time_seconds()
        self.create_timer(0.1, self.check_battery_state)
        self.get_logger().info("LED Client Node has been started")

    def get_time_seconds(self):
        sec, nano = self.get_clock().now().seconds_nanoseconds()
        return sec + nano / 1e9

    def check_battery_state(self):
        time_now = self.get_time_seconds()
        time_diff = time_now - self.last_time
        changed_state = False
        if self.battery_full and time_diff > 4.0:
            self.battery_full = False
            changed_state = True
            self.last_time = time_now
        elif not self.battery_full and time_diff > 6.0:
            self.battery_full = True
            changed_state = True
            self.last_time = time_now
        if changed_state:
            self.call_set_led(self.led_to_set, self.battery_full)

    def call_set_led(self, led, battery_full):
        client = self.create_client(BatteryStatus, "set_led")
        while not client.wait_for_service(0.5):
            self.get_logger().warn("Client waiting for server...")

        request = BatteryStatus.Request()
        request.led = led
        request.state = not battery_full

        future = client.call_async(request=request)
        future.add_done_callback(partial(self.set_led_callback, led=led, state=not battery_full))

    def set_led_callback(self, future, led, state):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f"LED {led} was set {state}")
        except Exception as e:
            self.get_logger().error(e)


def main(args=None):
    rclpy.init(args=args)

    node = LEDClientNode()

    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: tutocpp_pkg/src/activity_04_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "gonca_interfaces/srv/battery_status.hpp"

class LEDClientNode : public rclcpp::Node
{
public:
    LEDClientNode() : Node("LED_client_node"), ledToSet(3), batteryFull(true), changedState(false)
    {
        thread_ = std::thread(std::bind(&LEDClientNode::clientCallBack, this));
        RCLCPP_INFO(get_logger(), "LED client node has been started");
    }
    void clientCallBack()
    {
        auto client = create_client<gonca_interfaces::srv::BatteryStatus>("set_led");
        while (!client->wait_for_service(std::chrono::milliseconds(500)))
        {
            RCLCPP_WARN(get_logger(), "Client waiting for server...");
        }
        lastTime = std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch());
        while (true)
        {
            checkBatteryState();
            if (changedState)
            {
                changedState = false;
                auto request = std::make_shared<gonca_interfaces::srv::BatteryStatus::Request>();
                request->led = ledToSet;
                request->state = !batteryFull;

                auto future = client->async_send_request(request);
                try
                {
                    auto response = future.get();
                    if (response->success)
                        RCLCPP_INFO(get_logger(), "LED %ld was set %d", static_cast<long>(request->led), request->state);
                    else
                        RCLCPP_ERROR(get_logger(), "Error on setting LED");
                }
                catch (const std::exception &e)
                {
                    RCLCPP_ERROR(get_logger(), e.what());
                }
            }
        }
    }

private:
    int ledToSet;
    bool batteryFull, changedState;
    std::chrono::seconds lastTime;
    std::thread thread_;
    rclcpp::TimerBase::SharedPtr timer_;

    void checkBatteryState()
    {
        auto timeNow = std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::system_clock::now().time_since_epoch());
        auto timeDiff = timeNow - lastTime;
        if (batteryFull && timeDiff >= std::chrono::seconds(4))
        {
            lastTime = timeNow;
            batteryFull = false;
            changedState = true;
        }
        else if (!batteryFull && timeDiff >= std::chrono::seconds(6))
        {
            lastTime = timeNow;
            batteryFull = true;
            changedState = true;
        }
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<LEDClientNode>();

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}